src/cmd/package.go

```
@@ -336,20 +342,76 @@ func (o *packageMirrorResourcesOptions) run(cmd *cobra.Command, args []string) (
-		OCIConcurrency:        config.CommonOptions.OCIConcurrency,
-		PlainHTTP:             config.CommonOptions.PlainHTTP,
-		InsecureSkipTLSVerify: config.CommonOptions.InsecureSkipTLSVerify,
-	}
-	err = packager2.Mirror(ctx, mirrorOpt)
-	if err != nil {
-		return err
+	images, repos := 0, 0
+	// Let's count the images and repos in the package
+	for _, component := range pkgLayout.Pkg.Components {
+		images += len(component.Images)
+		repos += len(component.Repos)
+	}
+	logger.From(ctx).Debug("Package contains images and repos", "images", images, "repos", repos)
+
+	// We don't yet know if the targets are internal or external
+	c, _ := cluster.New(ctx) //nolint:errcheck
+
+	if o.imagesOnly {
+		if images == 0 {
+			return fmt.Errorf("no images found in package to mirror")
+		}
+		logger.From(ctx).Info("Mirroring images", "images", images)
+		if pkgConfig.InitOpts.RegistryInfo.Address == "" {
+			// if empty flag & zarf state available - execute
+			// otherwise return error
+			state, err := c.LoadState(ctx)
+			if err != nil {
+				return fmt.Errorf("no registry URL provided and no zarf state found")
+			}
+			logger.From(ctx).Debug("No registry URL provided, using zarf state", "address", state.RegistryInfo.Address)
+			pkgConfig.InitOpts.RegistryInfo = state.RegistryInfo
+		}
+		mirrorOpt := packager2.MirrorOptions{
+			Cluster:               c,
+			PkgLayout:             pkgLayout,
+			RegistryInfo:          pkgConfig.InitOpts.RegistryInfo,
+			NoImageChecksum:       pkgConfig.MirrorOpts.NoImgChecksum,
+			Retries:               pkgConfig.PkgOpts.Retries,
+			OCIConcurrency:        config.CommonOptions.OCIConcurrency,
+			PlainHTTP:             config.CommonOptions.PlainHTTP,
+			InsecureSkipTLSVerify: config.CommonOptions.InsecureSkipTLSVerify,
+		}
+		err = packager2.MirrorImages(ctx, mirrorOpt)
+		if err != nil {
+			return err
+		}
+	}
+
+	if o.reposOnly {
+		if repos == 0 {
+			return fmt.Errorf("no repos found in package to mirror")
+		}
```

This is a breaking change as if a user runs `zarf package mirror-resources` without either the `--images` or `--repos` flag and their package only has either images or repos then the command would fail. As a user I'd expect the command to silently succeed.

---

Added a `warn` log for each scenario but otherwise still a successful exit code

---

Maybe `Info` would be more appropriate? Nothing went wrong per say. Or I feel `Warn` would be appropriate if we only logged when a user used either the `--repos` or `--images` flag and they didn't have the specified resource

---

style guide would help here. There can be a lot in logging and `Info` might obfuscate an unintentional result? versus `Warn`

```bash
dev@dev:~/work/zarf$ ./build/zarf package mirror-resources zarf-package-git-data-amd64-0.0.1.tar.zst
2025-05-02 18:15:28 INF using config file location=/home/dev/work/zarf/zarf-config.toml
2025-05-02 18:15:28 WRN No images found in package to mirror
2025-05-02 18:15:28 INF Mirroring repos repos=9
2025-05-02 18:15:28 INF pushing repository to server repo=https://github.com/zarf-dev/zarf-public-test.git server=http://127.0.0.1:39711
2025-05-02 18:15:28 INF pushing repository to server repo=https://dev.azure.com/defenseunicorns/zarf-public-test/_git/zarf-public-test server=http://127.0.0.1:46253
2025-05-02 18:15:28 INF pushing repository to server repo=https://github.com/zarf-dev/zarf-public-test.git@v0.0.1 server=http://127.0.0.1:39859
2025-05-02 18:15:28 INF pushing repository to server repo=https://github.com/zarf-dev/zarf-public-test.git@refs/tags/v0.0.1 server=http://127.0.0.1:37995
2025-05-02 18:15:28 INF pushing repository to server repo=https://dev.azure.com/defenseunicorns/zarf-public-test/_git/zarf-public-test@v0.0.1 server=http://127.0.0.1:39677
2025-05-02 18:15:29 INF pushing repository to server repo=https://github.com/zarf-dev/zarf-public-test.git@refs/heads/dragons server=http://127.0.0.1:34441
2025-05-02 18:15:29 INF pushing repository to server repo=https://dev.azure.com/defenseunicorns/zarf-public-test/_git/zarf-public-test@refs/heads/dragons server=http://127.0.0.1:33961
2025-05-02 18:15:29 INF pushing repository to server repo=https://github.com/zarf-dev/zarf-public-test.git@01a23218923f24194133b5eb11268cf8d73ff1bb server=http://127.0.0.1:45897
2025-05-02 18:15:30 INF pushing repository to server repo=https://dev.azure.com/defenseunicorns/zarf-public-test/_git/zarf-public-test@01a23218923f24194133b5eb11268cf8d73ff1bb server=http://127.0.0.1:42151
```

More than willing to change it over to `Info`

---

I can see the argument for it either way, and won't block on it but I feel it's very unlikely that a user would want to mirror resources in a package and not realize the package doesn't have the resource they want to mirror. If they're running this command it's likely with a specific purpose in mind. 

Side note, the warning is currently always issued if the package has no repos, even if only `--images` was used.

---

I'll advocate for warn here but agree that the log presence is a rough edge that we can solve with a quick additional conditional.